commit 735a0045f9ea8372bcf3e599cbd3aa891a216b26
Author: Goodstein, Mordechay <mordechay.goodstein@intel.com>
Date:   Sun Jan 15 16:00:12 2017 +0200

    iwlwifi: mvm: avoid race condition in ADD_STA.
    
    The race happens when we send ADD_STA(auth->assoc) -> LQ_CMD
    between the commands the FW sometimes loses the medium for AUX, and
    sends a ndp to the AP and the flow becomes, ADD_STA -> send ndp -> LQ_CMD
    the problem is that there's no rates yet defined for sending the ndp and
    FW generates an assert.
    
    The fix: change the order of the commands to LQ_CMD -> ADD_STA
    
    Signed-off-by: Mordechay Goodstein <mordechay.goodstein@intel.com>
    Signed-off-by: Luca Coelho <luciano.coelho@intel.com>

diff --git a/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c b/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
index 7253b92792bf..d37b1695c64e 100644
--- a/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
+++ b/drivers/net/wireless/intel/iwlwifi/mvm/mac80211.c
@@ -2627,11 +2627,10 @@ static int iwl_mvm_mac_sta_state(struct ieee80211_hw *hw,
 			mvmvif->ap_assoc_sta_count++;
 			iwl_mvm_mac_ctxt_changed(mvm, vif, false, NULL);
 		}
+
+		iwl_mvm_rs_rate_init(mvm, sta, mvmvif->phy_ctxt->channel->band,
+				     true);
 		ret = iwl_mvm_update_sta(mvm, vif, sta);
-		if (ret == 0)
-			iwl_mvm_rs_rate_init(mvm, sta,
-					     mvmvif->phy_ctxt->channel->band,
-					     true);
 	} else if (old_state == IEEE80211_STA_ASSOC &&
 		   new_state == IEEE80211_STA_AUTHORIZED) {